zerver/migrations/0436_realmauthenticationmethods.py

```diff
@@ -0,0 +1,52 @@
+# Generated by Django 4.2 on 2023-04-13 23:45
+
+import django.db.models.deletion
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def fill_RealmAuthenticationMethod_data(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    RealmAuthenticationMethod = apps.get_model("zerver", "RealmAuthenticationMethod")
+    for realm in Realm.objects.order_by("id"):
+        rows_to_create = []
+        for key, value in realm.authentication_methods.iteritems():
+            if value:
+                rows_to_create.append(RealmAuthenticationMethod(name=key, realm_id=realm.id))
+        RealmAuthenticationMethod.objects.bulk_create(rows_to_create)
```

Comment:
Why are we calling `bulk_create` inside a loop here? This adds a database round trip per realm, and caused a stall of several minutes during deployment. I don’t see any reason we couldn’t have done this efficiently:

```diff
+    rows_to_create = []
     for realm in Realm.objects.order_by("id"):
-        rows_to_create = []
         for key, value in realm.authentication_methods.iteritems():
             if value:
                 rows_to_create.append(RealmAuthenticationMethod(name=key, realm_id=realm.id))
-        RealmAuthenticationMethod.objects.bulk_create(rows_to_create)
+    RealmAuthenticationMethod.objects.bulk_create(rows_to_create)
```

or efficiently and more Pythonically:

```python
    RealmAuthenticationMethod.objects.bulk_create(
        RealmAuthenticationMethod(name=key, realm_id=realm.id)
        for realm in Realm.objects.order_by("id")
        for key, value in realm.authentication_methods.iteritems()
        if value
    )
```

If there are nonobvious reasons to write SQL queries inside a Python loop (e.g. sometimes there are table locking considerations, though I don’t see them applying here), that should be considered and documented with a comment—but it should never be the *first* strategy we reach for.

Reply:
Yeah, it was silly to think that doing a trip per-realm was gonna be fine on Cloud. Opened https://github.com/zulip/zulip/pull/25420